src/ui/mainwindow.cpp

```diff
@@ -2217,15 +2221,33 @@ void MainWindow::AddStreamAccepted() {
 void MainWindow::OpenRipCDDialog() {
 #ifdef HAVE_AUDIOCD
   if (!rip_cd_dialog_) {
-    rip_cd_dialog_.reset(new RipCDDialog);
-  }
-  if (rip_cd_dialog_->CheckCDIOIsValid()) {
-    rip_cd_dialog_->show();
-  } else {
-    QMessageBox cdio_fail(QMessageBox::Critical, tr("Error"),
-                          tr("Failed reading CD drive"));
-    cdio_fail.exec();
+    QList<DeviceInfo*> cd_device_infos =
+        app_->device_manager()->FindDeviceByUrlSchemes(
+            CddaDevice::url_schemes());
+    std::shared_ptr<ConnectedDevice> device;
+    if (!cd_device_infos.isEmpty()) {
+      DeviceInfo* device_info =
+          cd_device_infos[0];  // TODO: currently just picking the first cd
+                               // drive; what if there are several?
+      device = app_->device_manager()->Connect(
+          device_info);  // connect the device; will do nothing if already
+                         // connected
+    }
```

Comment:
Could we move all of this into the ripper dialog? This seems like something the mainwindow shouldn't need to know about. Also, if there are multiple CD devices, you could have a list of devices in the dialog.

Reply:
Done.